src/object.d

```diff
@@ -38,6 +38,68 @@ alias dstring = immutable(dchar)[];
 
 version (D_ObjectiveC) public import core.attribute : selector;
 
+// Size used to store the TypeInfo at the end of an allocation for structs that
+// have a destructor
+private size_t structTypeInfoSize(T)() pure nothrow @nogc
+{
+    // can't use because it's in rt.lifetime
+    //if (!callStructDtorsDuringGC)
+        //return 0;
+
+    static if (is(T == struct) && is(typeof({ T t; t.__xdtor; })))
+        return size_t.sizeof;
+    return 0;
+}
+
+/*
+ * Allocate an uninitialized non-array item.
+ * This is an optimization to avoid things needed for arrays like the __arrayPad(size).
+ */
+private T* ___d_newitemU(T)()
+{
+    import core.memory;
+    import core.internal.traits : Unqual;
+
+    auto ti = typeid(Unqual!T);
+    auto flags = !(ti.flags & 1) ? GC.BlkAttr.NO_SCAN : 0;
```

Comment:
I thought the whole point was to avoid typeid? Surely you should be able to use a compiler trait for `ti.flags & 1`.

Reply:
Otherwise one can be added.

Reply:
I haven't found a compiler trait. Maybe I missed it. Anyone know about one?

Reply:
No, not that I can see. But add one. I think it's currently handled here: https://github.com/dlang/dmd/blob/bf26fb4ffc173f016eae9c241bcfd5a88da92957/src/dmd/toobj.d#L1291-L1327.

Reply:
I didn't know what `(ti.flags & 1)` meant, so I had to go spelunking in DMD source code to see what it means.  Assuming I'm reading the DMD source code correctly, it's `isCOMclass`.  At a minimum maybe you could do something like `enum COMClass = 1` to save the next poor soul the trouble.

Second, according to https://dlang.org/spec/interface.html#com-interfaces does that flag mean that it inherits from `core.stdc.windows.com.IUnknown`?  Is there an existing trait that can check for that?  Or can you use `__traits(compiles,...)` to achieve the desired result.

Reply:
It's really bad: https://github.com/dlang/druntime/blob/4ad638f61a9b4a98d8ed6eb9f9429c0ef6afc8e3/src/object.d#L1313-L1315

Especially as there is this easily confusing beauty (used for `TypeInfo_Class.m_flags`) you referred to: https://github.com/dlang/druntime/blob/4ad638f61a9b4a98d8ed6eb9f9429c0ef6afc8e3/src/object.d#L1969-L1980

Reply:
https://github.com/dlang/druntime/blob/4ad638f61a9b4a98d8ed6eb9f9429c0ef6afc8e3/src/object.d#L2194

Now I'm thoroughly confused.

Reply:
This one is for structs, not classes. TypeInfo_Class.flags() always returns 1 as TypeInfo_Class tries to describe both the reference and the instance of a class at the same time (which is causing trouble elsewhere, too, e.g. when trying to specify the type of memory for the precise GC).

Reply:
> Is there an existing trait that can check for that? Or can you use __traits(compiles,...) to achieve the desired result?

I guess `is(T : IUnknown)` can be used.